### PR TITLE
Detect 2019 modules and allow to install 2019, skipping dmg install for now

### PR DESCRIPTION
**Pull request by lacostej**
Definitively WIP with Windows not supported at all.

The approach is to use the `modules.asset` files to detect Playback Engines. @niezbop any better idea on how to support this?

I will also need some help to test this on Windows at some point. Maybe @BastianBlokland would like to test it on Linux?

### CODE REVIEW

**Comment by BastianBlokland** [comment]
Thanks, yeah should be able to test it out in a few hours.

**Comment by BastianBlokland** [comment]
Tried out the branch however there are a few issues still.

In line 261 of `installation.rb` it calls `IvyPlaybackEngineUtils.list_module_configs` and the line after it aborts if no package as found. Currently that doesn't go into the `modules.asset` path. Also 2 lines below that it tries to get the version from the ivy file.

If i change to `ModulePlaybackEngineUtils.list_module_configs` and hardcode the version then it work perfectly.

So i guess the biggest question remaining is how to get the version number.

**Comment by BastianBlokland** [comment]
Did a grep for it and only get binary files 😢
```
/opt/unity-editor-2019.2.13f1  grep -R "2019.2.13f1" *                                                                    ✔  978  20:53:07
Binary file Editor/Unity matches
Binary file Editor/Data/Resources/unity_builtin_extra matches
Binary file Editor/Data/Resources/unity editor resources matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.2d-3.2.3/SpriteAtlasDatabase.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.2d-3.2.3/SceneVisibilityState.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.2d-3.2.3/ScriptMapper matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.2d-3.2.3/InspectorExpandedItems.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.2d-3.2.3/AnnotationManager matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.2d-3.2.3/BuildSettings.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.2d-3.2.3/assetDatabase3 matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.2d-3.2.3/MonoManager.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.2d-3.2.3/EditorUserBuildSettings.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.lightweight-6.9.1/SpriteAtlasDatabase.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.lightweight-6.9.1/SceneVisibilityState.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.lightweight-6.9.1/ScriptMapper matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.lightweight-6.9.1/InspectorExpandedItems.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.lightweight-6.9.1/AnnotationManager matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.lightweight-6.9.1/BuildSettings.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.lightweight-6.9.1/assetDatabase3 matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.lightweight-6.9.1/MonoManager.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.lightweight-6.9.1/EditorUserBuildSettings.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.hd-6.9.0-preview/SpriteAtlasDatabase.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.hd-6.9.0-preview/SceneVisibilityState.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.hd-6.9.0-preview/ScriptMapper matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.hd-6.9.0-preview/InspectorExpandedItems.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.hd-6.9.0-preview/AnnotationManager matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.hd-6.9.0-preview/BuildSettings.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.hd-6.9.0-preview/assetDatabase3 matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.hd-6.9.0-preview/MonoManager.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.hd-6.9.0-preview/EditorUserBuildSettings.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3d-3.1.4/SpriteAtlasDatabase.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3d-3.1.4/SceneVisibilityState.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3d-3.1.4/ScriptMapper matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3d-3.1.4/InspectorExpandedItems.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3d-3.1.4/AnnotationManager matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3d-3.1.4/BuildSettings.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3d-3.1.4/assetDatabase3 matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3d-3.1.4/MonoManager.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3d-3.1.4/EditorUserBuildSettings.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3dextra-3.2.4/SpriteAtlasDatabase.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3dextra-3.2.4/SceneVisibilityState.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3dextra-3.2.4/ScriptMapper matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3dextra-3.2.4/InspectorExpandedItems.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3dextra-3.2.4/AnnotationManager matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3dextra-3.2.4/BuildSettings.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3dextra-3.2.4/assetDatabase3 matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3dextra-3.2.4/MonoManager.asset matches
Binary file Editor/Data/Resources/PackageManager/ProjectTemplates/libcache/com.unity.template.3dextra-3.2.4/EditorUserBuildSettings.asset matches
Binary file Editor/Data/Tools/WebExtract matches
Binary file Editor/Data/Tools/binary2text matches
Binary file Editor/Data/PlaybackEngines/LinuxStandaloneSupport/Variations/linux64_headless_nondevelopment_mono/LinuxPlayer matches
Binary file Editor/Data/PlaybackEngines/LinuxStandaloneSupport/Variations/linux64_withgfx_development_mono/LinuxPlayer matches
Binary file Editor/Data/PlaybackEngines/LinuxStandaloneSupport/Variations/linux64_withgfx_nondevelopment_mono/LinuxPlayer matches
Binary file Editor/Data/PlaybackEngines/LinuxStandaloneSupport/Variations/linux64_headless_development_mono/LinuxPlayer matches
Binary file Editor/Data/Managed/UnityEditor.Graphs.dll matches
Binary file Editor/Data/UnityExtensions/Unity/UnityVR/Editor/UnityEditor.VR.dll matches
Binary file Editor/BugReporter/UnityBugReporter matches
```

**Comment by niezbop** [comment]
@lacostej I think that this is a solid alternative to the ivy file since it seems to be as present as it used to be so can rely on this. However, this is super lackluster and does not support a ton of the modules. For instance:
```
$ u3d available -u 2019.2.12f1 -p
Version 2019.2.12f1: https://download.unity3d.com/download_unity/b1a7e1fb4fa5/
Packages:
 - Unity
 - Mono
 - Visualstudio
 - Android
 - Ios
 - Appletv
 - Linux
 - Mac-il2cpp
 - Webgl
 - Windows-mono
 - Facebook-games
 - Lumin
 - Documentation
 - Language-ja
 - Language-ko
 - Language-zh-hans
 - Language-zh-hant
 - Android-sdk-ndk-tools
 - Android-sdk-platform-tools
 - Android-sdk-build-tools
 - Android-sdk-platforms
 - Android-ndk
 - Android-open-jdk
```
Many of those would not be covered by the `modules.asset` file, for instance all NDK/SDK tools and platforms, which are rather sizeable and large to download.

So I think that if we want to address this issue correctly we need to tackle those as well. Languages could be fairly easy: `Dir.glob('Localization/*').map {|path| File.basename path }.map {|path| "Localization-#{path.gsub('.po', '')}" }`, while the other may be a bit trickier.

**Comment by niezbop** [comment]
@BastianBlokland Thanks a bunch for your help on this!

I think that we shouldn't be too afraid of looking into binaries. We already do this for Windows to get the version number from the exe:
https://github.com/DragonBox/u3d/blob/e980dddca928ab101daaaee0db28e8d7e5f48759/lib/u3d/installation.rb#L327-L362
So I believe that we should try looking into this.

**Comment by BastianBlokland** [comment]
Did some digging on the Unity executable, unfortunately all the occurrences seem to be in the `rodata` section of the elf file:
```
(Numbers of the left are byte offsets into the file)
 grep -abPo '.{0,50}2019.2.13f1.{0,50}' Unity
87694023: tingsm_EditorVersionm_EditorVersionWithRevision2019.2.13f1 (e20f6c7e5017)UI/UI-DefaultETC1.shaderUI/UI-Def
87896891: cesResource previews-buildBuiltinPreviewsUnity 2019.2.13f11.0.0-alpha1.0.0-alpha.11.0.0-alpha-11.0.0-alp
88997909: lized file version. File: "%s". Expected version: 2019.2.13f1. Actual version: %s.Unable to load the number of
91410862: torRandSameAsRandNvalue >= 30.0fvalue <= 40.0f2019.2.13f1_e20f6c7e5017value >= 30value < 40value > 30va
93488722: ainCurlExecutor::SubmitCurlRequestUnityPlayer/2019.2.13f1 (UnityWebRequest/1.0, libcurl/7.52.0-DEV)U
93682693: ke Gecko) Chrome/37.0.2062.94 Safari/537.36 Unity/2019.2.13f1 (unity3d.com;launcheronUnityEventhttp://editor
```

Sections of the Unity executable:
```
readelf -S Unity
There are 42 section headers, starting at offset 0x128d4420:
Section Headers:
  [Nr] Name              Type             Address           Offset
       Size              EntSize          Flags  Link  Info  Align
  [ 0]                   NULL             0000000000000000  00000000
       0000000000000000  0000000000000000           0     0     0
  [ 1] .interp           PROGBITS         0000000000000238  00000238
       000000000000001c  0000000000000000   A       0     0     1
  [ 2] .note.ABI-tag     NOTE             0000000000000254  00000254
       0000000000000020  0000000000000000   A       0     0     4
  [ 3] .note.gnu.build-i NOTE             0000000000000274  00000274
       0000000000000024  0000000000000000   A       0     0     4
  [ 4] .hash             HASH             0000000000000298  00000298
       0000000000008a64  0000000000000004   A       5     0     8
  [ 5] .dynsym           DYNSYM           0000000000008d00  00008d00
       000000000001bde0  0000000000000018   A       6     3     8
  [ 6] .dynstr           STRTAB           0000000000024ae0  00024ae0
       000000000001a099  0000000000000000   A       0     0     1
  [ 7] .gnu.version      VERSYM           000000000003eb7a  0003eb7a
       0000000000002528  0000000000000002   A       5     0     2
  [ 8] .gnu.version_r    VERNEED          00000000000410a8  000410a8
       00000000000002d0  0000000000000000   A       6     9     8
  [ 9] .rela.dyn         RELA             0000000000041378  00041378
       00000000004e01e0  0000000000000018   A       5     0     8
  [10] .rela.plt         RELA             0000000000521558  00521558
       0000000000007d70  0000000000000018   A       5    12     8
  [11] .init             PROGBITS         00000000005292c8  005292c8
       000000000000001d  0000000000000000  AX       0     0     4
  [12] .plt              PROGBITS         00000000005292f0  005292f0
       00000000000053b0  0000000000000010  AX       0     0     16
  [13] .text             PROGBITS         000000000052e700  0052e700
       0000000004d41441  0000000000000000  AX       0     0     256
  [14] text_env          PROGBITS         000000000526fb50  0526fb50
       0000000000002723  0000000000000000  AX       0     0     16
  [15] .fini             PROGBITS         0000000005272274  05272274
       000000000000000e  0000000000000000  AX       0     0     4
  [16] .rodata           PROGBITS         0000000005273000  05273000
       00000000014d5aa0  0000000000000000   A       0     0     4096
  [17] .eh_frame_hdr     PROGBITS         0000000006748aa0  06748aa0
       0000000000195d3c  0000000000000000   A       0     0     4
  [18] .eh_frame         PROGBITS         00000000068de7e0  068de7e0
       0000000000a312b4  0000000000000000   A       0     0     8
  [19] .gcc_except_table PROGBITS         000000000730fa94  0730fa94
       00000000004090b3  0000000000000000   A       0     0     4
  [20] .tbss             NOBITS           0000000007918b50  07718b50
       0000000000000068  0000000000000000 WAT       0     0     16
  [21] .init_array       INIT_ARRAY       0000000007918b50  07718b50
       0000000000000370  0000000000000000  WA       0     0     8
  [22] .ctors            PROGBITS         0000000007918ec0  07718ec0
       0000000000001c70  0000000000000000  WA       0     0     8
  [23] .dtors            PROGBITS         000000000791ab30  0771ab30
       0000000000000010  0000000000000000  WA       0     0     8
  [24] .jcr              PROGBITS         000000000791ab40  0771ab40
       0000000000000008  0000000000000000  WA       0     0     8
  [25] .data.rel.ro      PROGBITS         000000000791ab60  0771ab60
       000000000025ea80  0000000000000000  WA       0     0     32
  [26] .dynamic          DYNAMIC          0000000007b795e0  079795e0
       0000000000000410  0000000000000010  WA       6     0     8
  [27] .got              PROGBITS         0000000007b799f0  079799f0
       0000000000000780  0000000000000008  WA       0     0     8
  [28] .got.plt          PROGBITS         0000000007b7a170  0797a170
       00000000000029e8  0000000000000008  WA       0     0     8
  [29] .data             PROGBITS         0000000007b7cb60  0797cb60
       0000000000053860  0000000000000000  WA       0     0     32
  [30] .bss              NOBITS           0000000007bd03c0  079d03c0
       0000000000240850  0000000000000000  WA       0     0     64
  [31] .comment          PROGBITS         0000000000000000  079d03c0
       00000000000000eb  0000000000000001  MS       0     0     1
  [32] .debug_aranges    PROGBITS         0000000000000000  079d04b0
       000000000040e130  0000000000000000           0     0     16
  [33] .debug_info       PROGBITS         0000000000000000  07dde5e0
       0000000004481a1b  0000000000000000           0     0     1
  [34] .debug_abbrev     PROGBITS         0000000000000000  0c25fffb
       000000000017676e  0000000000000000           0     0     1
  [35] .debug_line       PROGBITS         0000000000000000  0c3d6769
       0000000000752b7a  0000000000000000           0     0     1
  [36] .debug_str        PROGBITS         0000000000000000  0cb292e3
       00000000024b8a3c  0000000000000001  MS       0     0     1
  [37] .debug_loc        PROGBITS         0000000000000000  0efe1d1f
       000000000299b4df  0000000000000000           0     0     1
  [38] .debug_ranges     PROGBITS         0000000000000000  1197d1fe
       0000000000f57090  0000000000000000           0     0     1
  [39] .shstrtab         STRTAB           0000000000000000  128d428e
       000000000000018d  0000000000000000           0     0     1
  [40] .symtab           SYMTAB           0000000000000000  128d4ea0
       000000000081b238  0000000000000018          41   70006     8
  [41] .strtab           STRTAB           0000000000000000  130f00d8
       0000000001e35265  0000000000000000           0     0     1
Key to Flags:
  W (write), A (alloc), X (execute), M (merge), S (strings), I (info),
  L (link order), O (extra OS processing required), G (group), T (TLS),
  C (compressed), x (unknown), o (OS specific), E (exclude),
  l (large), p (processor specific)
```

Not sure how we would access those in any sane way.

**Comment by BastianBlokland** [comment]
Kinda crazy but the `Editor/Data/Resources/unity_builtin_extra` file might be useable.

Seems to be present across a bunch of unity versions and has the version number at a fixed offset into the file:
```
od -j 20 -N 11 -a unity-editor-2017.4.10f1/Editor/Data/Resources/unity_builtin_extra
0000024   2   0   1   7   .   4   .   1   0   f   1

od -j 20 -N 10 -a unity-editor-2018.2.5f1/Editor/Data/Resources/unity_builtin_extra
0000024   2   0   1   8   .   2   .   5   f   1

od -j 20 -N 10 -a unity-editor-2019.1.0f2/Editor/Data/Resources/unity_builtin_extra
0000024   2   0   1   9   .   1   .   0   f   2
```

**Comment by bdominguez** [comment]
Any updates on this?

**Comment by lacostej** [comment]
> Any updates on this?

got delayed, and January was mostly offtime. Resuming this now.

**Comment by lacostej** [comment]
> @lacostej I think that this is a solid alternative to the ivy file since it seems to be as present as it used to be so can rely on this. However, this is super lackluster and does not support a ton of the modules. For instance:
> 
> ```
> $ u3d available -u 2019.2.12f1 -p
> Version 2019.2.12f1: https://download.unity3d.com/download_unity/b1a7e1fb4fa5/
> Packages:
>  - Unity
>  - Mono
>  - Visualstudio
>  - Android
>  - Ios
>  - Appletv
>  - Linux
>  - Mac-il2cpp
>  - Webgl
>  - Windows-mono
>  - Facebook-games
>  - Lumin
>  - Documentation
>  - Language-ja
>  - Language-ko
>  - Language-zh-hans
>  - Language-zh-hant
>  - Android-sdk-ndk-tools
>  - Android-sdk-platform-tools
>  - Android-sdk-build-tools
>  - Android-sdk-platforms
>  - Android-ndk
>  - Android-open-jdk
> ```
> 
> Many of those would not be covered by the `modules.asset` file, for instance all NDK/SDK tools and platforms, which are rather sizeable and large to download.
> 
> So I think that if we want to address this issue correctly we need to tackle those as well. Languages could be fairly easy: `Dir.glob('Localization/*').map {|path| File.basename path }.map {|path| "Localization-#{path.gsub('.po', '')}" }`, while the other may be a bit trickier.

According to https://forum.unity.com/threads/manual-install-sdk-ndk-tools-in-unity-2019-1-0f2.663907/ NDK goes under PlaybackEngines as well.

**Comment by lacostej** [comment]
This doesn't solve the reinstalll of the modules yet, but it at least makes  installing 2019 functional. Let's use it and release a new version with this, and I'll implement the #394 support after that. WDYT?

**Comment by bdominguez** [comment]
In our case we are interested in the original purpose of this pull request: to be able to correctly detect the installed modules.

We're wasting a lot of time doing unnecessary reinstallations through Jenkins.

I agree with the approach as long as it does not take too long to solve the main problem.

**Comment by lacostej** [comment]
@bdominguez I heard you! Just a bit too tight for me to deal with this this week. It's next in pipeline.

File: lib/u3d.rb [hunk 1]
```diff
@@ -48,4 +48,15 @@
 module U3d
   Helper = U3dCore::Helper
   UI = U3dCore::UI
+
+  def self.const_missing(const_name)
+    deprecated = {
+      PlaybackEngineUtils: IvyPlaybackEngineUtils,
+      INIParser: INIModulesParser
+    }
```
Comment (niezbop) [inline lib/u3d.rb]: I would move this out of the method, maybe to a constant like DEPRECATED_CLASSES or something

Comment (lacostej) [reply lib/u3d.rb]: Will do in a future cleanup.

**Pull request merged**